### PR TITLE
MSAL Node - HttpClient Proxy Improvements

### PR DESCRIPTION
**Pull request by Robbie-Microsoft**
Added `customAgent` to msal-node system options so custom http(s) agents can be used by the HttpClient. This allows devs to use agents created by third party libraries (useful for proxy support).

Added support for authenticated proxy.

NOTE: faq.md was automatically linted. The only change is on line 75.

### CODE REVIEW

File: lib/msal-node/src/config/Configuration.ts [hunk 1]
```diff
@@ -145,6 +146,7 @@ const DEFAULT_SYSTEM_OPTIONS: Required<NodeSystemOptions> = {
     loggerOptions: DEFAULT_LOGGER_OPTIONS,
     networkClient: new HttpClient(),
     proxyUrl: Constants.EMPTY_STRING,
+    customAgent: new http.Agent() || new https.Agent(),
```
Comment (bgavrilMS) [inline lib/msal-node/src/config/Configuration.ts]: What does it mean? Doesn't this just resolve to `new http.Agent()` ?

File: lib/msal-node/src/network/HttpClient.ts [hunk 2]
```diff
@@ -72,6 +78,7 @@ export class HttpClient implements INetworkModule {
                 this.proxyUrl,
                 HttpMethod.POST,
                 options,
+                this.customAgent as http.Agent,
```
Comment (bgavrilMS) [inline lib/msal-node/src/network/HttpClient.ts]: What if the customer configured https.Agent? Should you not throw an exception with a good error message explaining how to configure these?

File: lib/msal-node/src/network/HttpClient.ts [hunk 3]
```diff
@@ -80,6 +87,7 @@ export class HttpClient implements INetworkModule {
                 url,
                 HttpMethod.POST,
                 options,
+                this.customAgent as https.Agent,
```
Comment (bgavrilMS) [inline lib/msal-node/src/network/HttpClient.ts]: How will this work with MSI? Calls to MSI are done over http.

I don't expect anyone to define a proxy for calling MSI. That can be unsupported scenario.

**Review by bgavrilMS — CHANGES_REQUESTED** [review]
I understand this is not unit testable, but please explain how you manually tested this?

File: lib/msal-node/src/network/HttpClient.ts (hunk 2 above)
Comment (Robbie-Microsoft) [reply lib/msal-node/src/network/HttpClient.ts]: I made some improvements to address this.

Take a look at this [stack overflow](https://stackoverflow.com/questions/64117101/what-is-the-role-of-https-agent-in-node) post. 

> https.Agent internally calls and constructs with http.Agent

I interpret this as meaning either one can be used. Look at my improvements and let me know your thoughts.

File: lib/msal-node/src/network/HttpClient.ts (hunk 3 above)
Comment (Robbie-Microsoft) [reply lib/msal-node/src/network/HttpClient.ts]: My improvements should address this.

File: lib/msal-node/src/config/Configuration.ts (hunk 1 above)
Comment (Robbie-Microsoft) [reply lib/msal-node/src/config/Configuration.ts]: My improvements should address this.

**Comment by Robbie-Microsoft** [comment]
> I understand this is not unit testable, but please explain how you manually tested this?

I'll be testing this shortly on my Windows laptop with Fiddler, to simulate a proxy. I'm also going to to ask the user who filed the linked issue to test this.

**Pull request closed without merging**